test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mininode import P2PInterface, P2PDataStore
+from test_framework.util import assert_equal, wait_until
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
+
+import time
+
+
+""" Test node eviction logic
+
+When the number of peers has reached the limit of maximum connections,
+the next connecting inbound peer will trigger the eviction mechanism.
+We cannot currently test the parts of the eviction logic that are based on
+address/netgroup since in the current framework, all peers are connecting from
+the same local address. See Issue #14210 for more info.
+
+Therefore, this test is limited to the remaining protection criteria:
+1.) Protect 4 nodes having sent us a block most recently
+2.) Protect 4 nodes having sent us a transaction most recently
+3.) Protect 8 nodes with the smallest minimum ping
+"""
+
+class SlowP2PDataStore(P2PDataStore):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+class SlowP2PInterface(P2PInterface):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+
+class P2PEvict(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        # The choice of maxconnections results in a maximum of 21 inbound connections
+        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
+        self.extra_args = [['-maxconnections=32']]
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def run_test(self):
+        protected = set()
+        node = self.nodes[0]
+        node.generatetoaddress(101, node.get_deterministic_priv_key().address)
+
+        #The first 4 peers send us a block, protecting them from eviction
+        for i in range(4):
+            blockpeer = node.add_p2p_connection(SlowP2PDataStore())
+            blockpeer.sync_with_ping(timeout=5)
+            best_block = node.getbestblockhash()
+            tip = int(best_block, 16)
+            best_block_time = node.getblock(best_block)['time']
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            blockpeer.send_blocks_and_test([block], node, success=True)
+            protected.add(len(node.p2ps))
```

Comment:
Although I understand how it works, this construction takes a little bit of cognitive effort.

Fine for this one, but I suggest using something explicit for the next PRs. Simple `i` would work here, and below, well, you'd have to introduce `i + N(4) + M(5)` or something, but I think it would be easier to comprehend.

Reply:
I introduced a variable `current_peer`, hopefully that makes it easier to read.